src/test/java/org/apache/sysml/test/gpu/NeuralNetworkOpTests.java

```diff
@@ -0,0 +1,490 @@
+
+		for (long N : Nlst) {
+			for (long C : Clst) {
+				for (long H : Hlst) {
+					for (long W : Wlst) {
+						for (long K : Klst) {
+							for (long R : Rlst) {
+								for (long S : Slst) {
+									for (long strideH : strideHeightLst) {
+										for (long strideW : strideWidthLst) {
+											for (long padH : padHeightLst) {
+												for (long padW : padWidthLst) {
+													for (double sparsity : sparsitylst) {
+
+														// Make sure ops fit in GPU memory and within constraints of cudnn
+														long imageSize = N * C * H * W * 8l;
+														if (imageSize > MAX_OP_SIZE)  // image size
+															continue;
+														long filterSize = K * C * R * S * 8l;
+														if (filterSize > MAX_OP_SIZE)  // filter size
+															continue;
+														// filter is smaller than image + padding
+														if (R > (H + padH) || S > (W + padW))
+															continue;
+
+														int P = (int) ConvolutionUtils.getP(H, R, strideH, padH);
+														int Q = (int) ConvolutionUtils.getQ(W, S, strideW, padW);
+
+														long doutSize = N * K * P * Q * 8l;
+														if (doutSize > MAX_OP_SIZE) // dout/output size
+															continue;
+
+														double imageSizeInMB = imageSize / (1024.0 * 1024.0);
+														double filterSizeInMB = filterSize / (1024.0 * 1024.0);
+														double doutSizeInMB = doutSize / (1024.0 * 1024.0);
+														System.out
+																.format("conv2d_backward_filter, image[%d,%d,%d,%d](%.1fMB), filter[%d,%d,%d,%d](%.1f), dout[%d,%d,%d,%d](%.1fMB), stride[%d,%d], padding[%d,%d]",
+																		N, C, H, W, imageSizeInMB, N, C, R, S,
+																		filterSizeInMB, N, K, P, Q, doutSizeInMB,
+																		strideH, strideW, padH, padW);
+														Matrix image = generateInputMatrix(spark, (int) N,
+																(int) (C * H * W), sparsity, seed);
+														Matrix filter = generateInputMatrix(spark, (int) K,
+																(int) (C * R * S), sparsity, seed);
+														HashMap<String, Object> inputs = new HashMap<>();
+														inputs.put("N", N);
+														inputs.put("C", C);
+														inputs.put("H", H);
+														inputs.put("W", W);
+														inputs.put("K", K);
+														inputs.put("R", R);
+														inputs.put("S", S);
+														inputs.put("strideH", strideH);
+														inputs.put("strideW", strideW);
+														inputs.put("padH", padH);
+														inputs.put("padW", padW);
+														inputs.put("image", image);
+														inputs.put("filter", filter);
+														List<Object> outCPU = runOnCPU(spark, scriptStr, inputs,
+																Arrays.asList("O"));
+														List<Object> outGPU = runOnGPU(spark, scriptStr, inputs,
+																Arrays.asList("O"));
+														assertHeavyHitterPresent("gpu_conv2d");
+														assertEqualObjects(outCPU.get(0), outGPU.get(0));
```

Comment:
Is this testing absolute error difference or a relative error difference?  I would recommend using a relative error or otherwise the magnitude of the values being compared could mask an issue.  For example, if the two values are ~1e5, the difference between them ends up being ~1e6, and we use a difference threshold of 1e5, then the test would pass, but the difference is actually relatively huge, and likely indicative of an issue.